include/seqan3/alphabet/detail/alphabet_proxy.hpp

```diff
@@ -0,0 +1,292 @@
+ * details.
+ *
+ * This class ensures that the proxy itself also models seqan3::semi_alphabet_concept, seqan3::alphabet_concept,
+ * seqan3::quality_concept, seqan3::nucleotide_concept and/or seqan3::aminoacid_concept if the emulated type models
+ * these. This makes sure that function templates which accept the original, also accept the proxy. An exception
+ * are multi-layered compositions of alphabets where the proxy currently does not support access via `get`.
+ *
+ * ### Implementation notes
+ *
+ * The derived type needs to provide an `.impl()` member function that performs the changes in the underlying
+ * data structure.
+ *
+ * See seqan3::bitcompressed_vector or seqan3::cartesian_composition for examples of how this class is used.
+ */
+template <typename derived_type, typename alphabet_type>
+class alphabet_proxy : public alphabet_base<derived_type,
+                                            alphabet_size_v<alphabet_type>,
+                                            detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>>
+{
+private:
+    //!\brief Type of the base class.
+    using base_t =  alphabet_base<derived_type,
+                                  alphabet_size_v<alphabet_type>,
+                                  detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>>;
+
+    //!\brief Befriend the base type.
+    friend base_t;
+
+public:
+    // Import from base:
+    using base_t::value_size;
+    using base_t::to_rank;
+
+    /*!\name Member types
+     * \{
+     */
+    using rank_type  = underlying_rank_t<alphabet_type>;
+    using char_type  = detail::transformation_trait_or_t<underlying_char<alphabet_type>, void>;
+    using phred_type = detail::transformation_trait_or_t<underlying_phred<alphabet_type>, void>;
+    //!\}
+
+private:
+    //!\brief Never used, but required for valid definitions.
+    using char_type_virtual  = std::conditional_t<std::Same<char_type, void>, char, char_type>;
+    //!\brief Never used, but required for valid definitions.
+    using phred_type_virtual = std::conditional_t<std::Same<phred_type, void>, int8_t, phred_type>;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_proxy() : base_t{} {}
+    constexpr alphabet_proxy(alphabet_proxy const &) = default;
+    constexpr alphabet_proxy(alphabet_proxy &&) = default;
+    constexpr alphabet_proxy & operator=(alphabet_proxy const &) = default;
+    constexpr alphabet_proxy & operator=(alphabet_proxy &&) = default;
+    ~alphabet_proxy() = default;
+
+    //!\brief Construction from the emulated type.
+    constexpr alphabet_proxy(alphabet_type const a)
+    {
+        base_t::assign_rank(seqan3::to_rank(a));
+    }
+
+    //!\brief Assigment from the emulated type. This function triggers the specialisation in the derived_type.
+    constexpr derived_type & operator=(alphabet_type const & c) noexcept
```

Comment:
Is this noexecpt? Because your all your examples, like reference_proxy_type, don't have noexecpt

Reply:
They had `constexpr` which was sufficient, but I have now started adding `noexcept` everywhere, too.